### ISSUE TITLE
Stages functions got removed when deploying to different stage

### ISSUE THREAD

**Issue by berkaey** [serverless/serverless#7232]
# Bug Report

As title states, when I trying to deploy to `production`, `staging` functions got removed, and deploying to `staging` removes `production` functions. btw, API Gateway stages is kept.

`sls deploy` to deploy staging

`sls deploy --stage 'production'` to deploy production

## Description

1. What did you do?
Deploy to different stage
1. What happened?
Other stage got removed.
1. What should've happened?
Other stages should be kept.
1. What's the content of your `serverless.yml` file?

`service: xxx-api

plugins:
  - serverless-domain-manager
  - serverless-offline

provider:
  name: aws
  profile: jokerapi
  stage: ${opt:stage, 'staging'}
  region: ${opt:region, 'eu-central-1'}
  stackName: ${self:service}
  runtime: nodejs12.x

custom:
  domain:
    staging: api.staging.xxx.com
    production: api.xxx.com
  customDomain:
    domainName: ${self:custom.domain.${self:provider.stage}}
    stage: ${self:provider.stage}
    basePath: ""
    createRoute53Record: false

functions:
  v1:
    handler: handler.handler
    name: ${self:service}-${self:provider.stage}-v1
    events:
      - http:
          path: v1/{proxy+}
          method: any
          cors: true

  fallback:
    handler: handler.fallback
    name: ${self:service}-${self:provider.stage}-fallback
    events:
      - http:
          path: /{proxy+}
          method: any
          cors: true

  home:
    handler: handler.home
    name: ${self:service}-${self:provider.stage}-home
    events:
      - http:
          path: /
          method: any
          cors: true

resources:
  Resources:
    ApiGatewayRestApi:
      Type: AWS::ApiGateway::RestApi
      Properties:
        Name: ${self:service}
`

1. What's the output you get when you use the `SLS_DEBUG=*` environment variable (e.g. `SLS_DEBUG=* serverless deploy`)

No CLI error.

Similar or dependent issues:

- #12345

**Comment by berkaey** [comment]
btw, changing `  stackName: ${self:service} ` to `   stackName: ${self:service}-${custom:stage} ` does the trick but now I have to separate APIGateway, how can I force stacks to share same gateway?

**Comment by medikoo** [comment]
@berkaey by overriding the `stackName` you make _stage_ setting ineffective, as then same CF stack is used for any stage (so technically there's no stages).

> how can I force stacks to share same gateway?

See: https://serverless.com/framework/docs/providers/aws/events/apigateway/#easiest-and-cicd-friendly-example-of-using-shared-api-gateway-and-api-resources

**Comment by berkaey** [comment]
@medikoo yes I noticed that too, it works with this setup but could be better with shared API Gateway 👍
